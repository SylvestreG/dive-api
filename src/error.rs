use thiserror::Error;

#[derive(Error, Debug)]
pub enum DiveErr {
    #[error("diesel Error")]
    IoErr(#[from] diesel::result::Error),
    #[error("r2d2 Error")]
    R2d2Err(#[from] r2d2::Error),
    #[error("dotenv Error")]
    DotenvError(#[from] dotenvy::Error),
    #[error("uuid Erro")]
    UuidError(#[from] uuid::Error),
}
