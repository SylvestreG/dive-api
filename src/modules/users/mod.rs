mod fixtures;
mod models;
mod repositories;
pub mod services;
#[cfg(test)]
mod tests;
