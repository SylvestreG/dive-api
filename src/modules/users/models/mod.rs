pub mod level;
pub mod user;
pub mod user_level;
