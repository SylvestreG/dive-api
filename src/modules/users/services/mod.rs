pub mod level;
pub mod user;
