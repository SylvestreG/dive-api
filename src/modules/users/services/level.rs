use crate::error::DiveErr;
use crate::modules::users::{fixtures, models, repositories};
use diesel::r2d2::ConnectionManager;
use diesel::PgConnection;
use r2d2::Pool;
use uuid::Uuid;

pub struct Service {
    repository: repositories::level::Repository,
}

impl Service {
    #[must_use]
    pub fn new(con: Pool<ConnectionManager<PgConnection>>) -> Self {
        Service {
            repository: repositories::level::Repository::new(con),
        }
    }

    #[must_use]
    pub fn fixtures(&self) -> Box<fixtures::level::Fixtures> {
        Box::new(fixtures::level::Fixtures::new(self.repository.clone()))
    }

    pub fn find_by_uuid(&self, uuid: Uuid) -> Result<models::level::Entity, DiveErr> {
        // TODO ACL

        self.repository.get(uuid)
    }

    pub fn list_all_levels(&self) -> Result<Vec<models::level::Entity>, DiveErr> {
        // TODO ACL

        self.repository.all()
    }

    #[must_use]
    pub fn user_by_role(&self) -> Vec<models::level::Entity> {
        vec![]
    }
}
