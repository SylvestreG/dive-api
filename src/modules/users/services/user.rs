use crate::modules::users::{fixtures, models, repositories};
use diesel::r2d2::ConnectionManager;
use diesel::PgConnection;
use r2d2::Pool;
use uuid::Uuid;
use crate::error::DiveErr;

pub struct Service {
    repository: repositories::user::Repository,
}

impl Service {

    #[must_use]
    pub fn new(con: Pool<ConnectionManager<PgConnection>>) -> Self {
        Service {
            repository: repositories::user::Repository::new(con),
        }
    }

    #[must_use]
    pub fn fixtures(&self) -> Box<fixtures::user::Fixtures> {
        Box::new(fixtures::user::Fixtures::new(self.repository.clone()))
    }

    #[must_use]
    pub fn find_by_uuid(&self, _uuid: Uuid) -> models::user::Entity {
        todo!()
    }

    #[must_use]
    pub fn list_all_user(&self) -> Result<Vec<models::user::Entity>, DiveErr> {
        self.repository.all()
    }

    #[must_use]
    pub fn user_by_role(&self) -> Vec<models::user::Entity> {
        vec![]
    }
}
