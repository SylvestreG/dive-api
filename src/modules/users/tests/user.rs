use diesel::PgConnection;
use diesel::r2d2::ConnectionManager;
use dotenvy::dotenv;
use r2d2::Pool;
use test_context::{test_context, TestContext};
use crate::error::DiveErr;
use crate::modules::users::services;

struct UserTests {
    svc: services::user::Service,
}

impl TestContext for UserTests {
    fn setup() -> UserTests {
        dotenv().unwrap();

        let db_url = std::env::var("DATABASE_URL").unwrap();
        let manager =
            ConnectionManager::<PgConnection>::new(db_url);
        let pool = Pool::builder().build(manager).unwrap();


        UserTests { svc: services::user::Service::new(pool)}
    }

    fn teardown(self) {
        // Perform any teardown you wish.
    }
}

#[test_context(UserTests)]
#[test]
fn list_all_user(ctx: &mut UserTests) -> Result<(), DiveErr>{
    let mut users = ctx.svc.list_all_user()?;

    users.sort();
    insta::assert_json_snapshot!(users, {
        "[].id" => "[uuid]",
        "[].created_at" => "[date]",
        "[].updated_at" => "[date]",
    });
    Ok(())
}
