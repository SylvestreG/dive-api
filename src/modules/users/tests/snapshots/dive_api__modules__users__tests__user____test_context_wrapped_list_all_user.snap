---
source: src/modules/users/tests/user.rs
expression: users
---
[
  {
    "id": "5232de05-4495-4fec-a45f-18bbef3f8ec3",
    "firstname": "toto",
    "lastname": "tutu",
    "mail": "tutu@ds.net",
    "tel": "0343239320",
    "blocked_at": null,
    "blocked_by": null,
    "created_at": "[date]",
    "updated_at": "[date]"
  },
  {
    "id": "888607ea-887f-415a-89cd-20df3153e009",
    "firstname": "aiai",
    "lastname": "yuyu",
    "mail": "aiai@ds.net",
    "tel": "0663903521",
    "blocked_at": null,
    "blocked_by": null,
    "created_at": "[date]",
    "updated_at": "[date]"
  },
  {
    "id": "f339cf72-47c5-48f8-9ad1-ff1d93f6c2e6",
    "firstname": "yuyu",
    "lastname": "yiyi",
    "mail": "gallon@ponay.fr",
    "tel": "0343239320",
    "blocked_at": null,
    "blocked_by": null,
    "created_at": "[date]",
    "updated_at": "[date]"
  },
  {
    "id": "f633eb82-ff98-4e93-9264-a27c1bb54c87",
    "firstname": "titi",
    "lastname": "tata",
    "mail": "tata@ponay.fr",
    "tel": "0343239320",
    "blocked_at": null,
    "blocked_by": null,
    "created_at": "[date]",
    "updated_at": "[date]"
  }
]
